### DISCUSSION TITLE
Music/sound creation

### DISCUSSION THREAD

**Discussion by archgary** [cables-gl/cables_docs#430] [Answers & Questions] [answered]
Hi all,

How do you all usually create the real-time sound effect with Cables.gl?

I certainly know that I can link to DAW software with cables.gl, but is there any alternative way you are using for real-time sound creations? (link to external javascript library? or how do I create a synthesizer with Cables?)

The reasons I am asking are that:
1. I would like to reduce multiple software' data transmission.
2. If I would like to set up an installation remotely, there will be lesser technical/financial issues get involved (like, I cannot expect the curator to get the Ableton live for me).

Would like to know how you all work on this issue. 

Thank you and look forward to any of your suggestions. 

Best,
archgary

**Comment by bejucoo** [comment, marked as answer]
I think cables can work with [tone.js](https://tonejs.github.io/) and other libraries [via WebAudioAPI](https://cables.gl/docs/5_writing_ops/webaudio/webaudio) if that's what you mean. Other options could be [Pizzicato.js](https://alemangui.github.io/pizzicato/) and [WadJS](https://github.com/rserota/wad).

**Reply by archgary** [reply]
Hey @bejucoo, 

Thanks for the reply! 
Those are exactly something really helpful and I would like to start with. Excellent suggestions! I am going to try. 

Of course, still welcome those who have others' advice to post with this discussion. 

Best,
archgary

**Reply by archgary** [reply]
I tried to follow the step in [WebAudioAPI](https://cables.gl/docs/5_writing_ops/webaudio/webaudio), and it said that "I am not allowed to add the library to this OP". Is there any solution to this? Thank you!

**Reply by steam0r** [reply]
you need to create a new userop first or clone an existing one

**Reply by archgary** [reply]
@steam0r thanks for your quick response. 

I try to make a new CustomOp, but it doesn't work. While I clone an existing one, it works. 
But the next question is, I cannot find the Tone.js lib to add
<img width="1280" alt="chrome_VbI9HEc5dX" src="https://user-images.githubusercontent.com/18259890/207882712-77f014ce-de7a-4889-94f6-05de8ebd57d2.png">
Or do I mis-understand where to add the library? Please suggest, appreciate it!

**Reply by steam0r** [reply]
thats corelibs, use the dropdown below

![Bildschirm­foto 2022-12-15 um 15 17 21](https://user-images.githubusercontent.com/151271/207883996-5a80c6f4-6b23-4d25-8d8c-2919f5106a44.png)

on the ops: ignore "CustomOp", create new userops here:

![Bildschirm­foto 2022-12-15 um 15 18 15](https://user-images.githubusercontent.com/151271/207884131-48d2c42c-e6d9-4558-9425-2d36654a9f9e.png)

**Reply by archgary** [reply]
@steam0r , thank you, I need to drag up the output display window to see the lib option below. 
Thanks, I will try to play and see if I bump into any other questions!

**Reply by archgary** [reply]
@steam0r, Sorry to ask one more question here. 

When saying "In order to use tone.js there is one thing left to do: In your op-code add the following line below your **op-name definition,** it should look like this:"  But where exactly should I put those lines in the code? Is it like this?
<img width="1280" alt="chrome_ooIyWsSdYK" src="https://user-images.githubusercontent.com/18259890/207891823-67602411-8219-4fc9-8daf-519fab131cbd.png">

To be able to use Tone.js, I will have to dive into creating my own OP with coding, am I right?
And if I consider using other libraries, are there any requirement lines needed in the OP like "CABLES.WEBAUDIO.createAudioContext();
Tone.setContext(window.audioContext);" ?

Thank you so much for your answers and help.

Best,
archgary

**Reply by bejucoo** [reply]
Yes, I think that's where you should put the code, did it work?. For other libraries I am not sure, did you try? Also, I forgot to mention you can check the [WebAudio OPs](https://cables.gl/ops/Ops.WebAudio).
